pubsub/schemas/schemas_test.go

```diff
@@ -0,0 +1,349 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package schema is a tool to manage Google Cloud Pub/Sub schemas by using the Pub/Sub API.
+// See more about Google Cloud Pub/Sub at https://cloud.google.com/pubsub/docs/overview
+package schema
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/GoogleCloudPlatform/golang-samples/internal/testutil"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/uuid"
+	"google.golang.org/api/iterator"
+)
+
+const (
+	schemaPrefix  = "test-schema-"
+	avroFilePath  = "./resources/us-states.avsc"
+	protoFilePath = "./resources/us-states.proto"
+
+	topicPrefix = "test-topic-"
+	subPrefix   = "test-sub-"
+)
+
+// once guards cleanup related operations in setup. No need to set up and tear
+// down every time, so this speeds things up.
+var once sync.Once
+
+func setup(t *testing.T) *pubsub.Client {
+	ctx := context.Background()
+	tc := testutil.SystemTest(t)
+
+	client, err := pubsub.NewClient(ctx, tc.ProjectID)
+	if err != nil {
+		t.Fatalf("failed to create client: %v", err)
+	}
+
+	// Cleanup resources from the previous tests.
+	// This includes schemas, topics, and subscriptions.
+	once.Do(func() {
+		scs, err := listSchemas(io.Discard, tc.ProjectID)
+		if err != nil {
+			t.Fatalf("failed to list schemas: %v", err)
+		}
+		for _, sc := range scs {
+			schemaName := strings.Split(sc.Name, "/")
+			deleteSchema(io.Discard, tc.ProjectID, schemaName[len(schemaName)-1])
+		}
+
+		go func() {
```

Comment:
this seems like it could run late and kill resources under test due to late execution?  Should these goroutines be in a waitgroup or some other control structure?  once.Do() ensures they run only once, but it doesn't ensure they complete before returning.